lib/rules/media-feature-name-unit-allowed-list/index.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const validateOptions = require('../../utils/validateOptions');
+const { TokenType } = require('@csstools/css-tokenizer');
+const { isTokenNode } = require('@csstools/css-parser-algorithms');
+const {
+	isMediaFeaturePlain,
+	isMediaFeatureRange,
+	parse,
+} = require('@csstools/media-query-list-parser');
+const validateObjectWithArrayProps = require('../../utils/validateObjectWithArrayProps');
+const { isString } = require('../../utils/validateTypes');
+const matchesStringOrRegExp = require('../../utils/matchesStringOrRegExp');
+const flattenArray = require('../../utils/flattenArray');
+
+const ruleName = 'media-feature-name-unit-allowed-list';
+
+const messages = ruleMessages(ruleName, {
+	rejected: (unit, name) => `Unexpected unit "${unit}" for name "${name}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/media-feature-name-unit-allowed-list',
+};
+
+/** @type {import('stylelint').Rule} */
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: [validateObjectWithArrayProps(isString)],
+		});
+
+		if (!validOptions) {
+			return;
+		}
+
+		root.walkAtRules(/^media$/i, (atRule) => {
+			const mediaQueryList = parse(atRule.params);
+
+			mediaQueryList.forEach((mediaQuery) => {
+				mediaQuery.walk((entry) => {
+					if (!isMediaFeaturePlain(entry.node) && !isMediaFeatureRange(entry.node)) {
+						return;
+					}
+
+					const name = entry.node.name.toString();
+
+					const nameKey = Object.keys(primary).find((nameIdentifier) =>
+						matchesStringOrRegExp(name, nameIdentifier),
+					);
+
+					if (!nameKey) {
+						return;
+					}
+
+					const nameList = flattenArray(primary[nameKey]);
+
+					if (!nameList) {
+						return;
+					}
+
+					entry.node.walk((childEntry) => {
+						if (!isTokenNode(childEntry.node) || childEntry.node.value[0] !== TokenType.Dimension) {
```

Comment:
Should this rule warn when authors use values without unit or percentages?

Reply:
My initial gut feeling is no; there are many media features that only have valid values that have no unit/percentage (see: [video-dynamic-range](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/video-dynamic-range) for an example of a number-based one). I've added a few tests that address various cases this rule should skip; not sure if this is the preferred way to test for this.

Reply:
I agree, it shouldn't.
e.g. `@media screen and (min-aspect-ratio: 16/9) {)` is unit less

Reply:
I was thinking of these types of issues :

```css
@media (width <= 100) {} /* number is not a length */
@media (width <= 50%) {} /* percentage is not a length */
@media (min-aspect-ratio: 16px/9px) {} /* length values do not make up a ratio */
```

But maybe that is beyond the scope of this rule?

Reply:
Good point. I believe such problems should be detected by a rule to [avoid errors](https://stylelint.io/user-guide/rules#avoid-errors). Perhaps we need a rule like `media-feature-no-invalid-range`.

Reply:
Drafted https://github.com/stylelint/stylelint/issues/6560. I struggled a bit with both the scope and behaviour of the rule; comments welcome!